### PR TITLE
Fix duplicating integrations due to multiple Initialize calls from different AppDomains.

### PR DESCRIPTION
**Pull request by tonyredondo**
This PR includes:

- Fix duplicating integrations due to multiple AppDomains.
- Fix memory leak on NativeCallTargetDefinition disposal.

#### Before:

```
09/20/21 03:34:59.283 PM [14204|12664] [info] InitializeProfiler: received from managed side: 222 integrations.
09/20/21 03:34:59.285 PM [14204|7460] [info] Request ReJIT done for 6 methods
09/20/21 03:34:59.285 PM [14204|7460] [info] Request ReJIT done for 1 methods
09/20/21 03:34:59.286 PM [14204|7460] [info] Request ReJIT done for 12 methods
09/20/21 03:34:59.286 PM [14204|7460] [info] Request ReJIT done for 1 methods
09/20/21 03:34:59.286 PM [14204|7460] [info] Request ReJIT done for 4 methods
09/20/21 03:34:59.286 PM [14204|7460] [info] Request ReJIT done for 6 methods
09/20/21 03:34:59.287 PM [14204|12664] [info] InitializeProfiler: Total integrations in profiler: 444
09/20/21 03:34:59.389 PM [14204|7460] [info] Request ReJIT done for 2 methods
09/20/21 03:34:59.545 PM [14204|9984] [info] *** CallTarget_RewriterCallback() Finished: System.Net.HttpWebRequest.BeginGetRequestStream() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.HttpWebRequest_BeginGetRequestStream_Integration, Arguments=2]
09/20/21 03:34:59.553 PM [14204|7460] [info] Request ReJIT done for 8 methods
09/20/21 03:34:59.615 PM [14204|9984] [info] *** CallTarget_RewriterCallback() Finished: System.Net.WebRequest.GetResponseAsync() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.WebRequest_GetResponseAsync_Integration, Arguments=0]
09/20/21 03:34:59.625 PM [14204|9984] [info] *** CallTarget_RewriterCallback() Finished: System.Net.HttpWebRequest.BeginGetResponse() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.HttpWebRequest_BeginGetResponse_Integration, Arguments=2]
09/20/21 03:34:59.659 PM [14204|10988] [info] *** CallTarget_RewriterCallback() Finished: System.Net.HttpWebRequest.EndGetResponse() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.HttpWebRequest_EndGetResponse_Integration, Arguments=1]
09/20/21 03:34:59.751 PM [14204|12664] [info] AssemblyLoadFinished: Datadog.Trace.dll v1.28.6.0 matched profiler version v1.28.6.0
09/20/21 03:34:59.751 PM [14204|12664] [info] AssemblyLoadFinished: Datadog.Trace.dll was not loaded domain-neutral
09/20/21 03:34:59.778 PM [14204|7460] [info] Request ReJIT done for 24 methods
09/20/21 03:35:00.102 PM [14204|12664] [info] JITCompilationStarted: Startup hook registered in function_id=186865732 token=100664278 name=Xunit.AssemblyHelper..ctor(), assembly_name=xunit.runner.utility.net452 app_domain_id=196679768 domain_neutral=0
09/20/21 03:35:00.122 PM [14204|12664] [info] ModuleLoadFinished: Datadog.Trace.ClrProfiler.Managed.Loader loaded into AppDomain 196679768 Datadog.Trace.DuckTyping.Tests
09/20/21 03:35:00.248 PM [14204|12664] [info] AssemblyLoadFinished: Datadog.Trace.dll v1.28.6.0 matched profiler version v1.28.6.0
09/20/21 03:35:00.248 PM [14204|12664] [info] AssemblyLoadFinished: Datadog.Trace.dll was not loaded domain-neutral
09/20/21 03:35:00.266 PM [14204|7460] [info] Request ReJIT done for 12 methods
09/20/21 03:35:00.466 PM [14204|12664] [info] InitializeProfiler: received from managed side: 222 integrations.
09/20/21 03:35:00.469 PM [14204|7460] [info] Request ReJIT done for 6 methods
09/20/21 03:35:00.469 PM [14204|7460] [info] Request ReJIT done for 1 methods
09/20/21 03:35:00.470 PM [14204|7460] [info] Request ReJIT done for 12 methods
09/20/21 03:35:00.470 PM [14204|7460] [info] Request ReJIT done for 4 methods
09/20/21 03:35:00.471 PM [14204|7460] [info] Request ReJIT done for 12 methods
09/20/21 03:35:00.471 PM [14204|7460] [info] Request ReJIT done for 1 methods
09/20/21 03:35:00.471 PM [14204|7460] [info] Request ReJIT done for 4 methods
09/20/21 03:35:00.472 PM [14204|7460] [info] Request ReJIT done for 6 methods
09/20/21 03:35:00.472 PM [14204|7460] [info] Request ReJIT done for 1 methods
09/20/21 03:35:00.473 PM [14204|7460] [info] Request ReJIT done for 6 methods
09/20/21 03:35:00.473 PM [14204|12664] [info] InitializeProfiler: Total integrations in profiler: 666
```

#### After:

```
09/20/21 03:29:16.892 PM [10372|9972] [info] InitializeProfiler: received id: FFAFA5168C4F4718B40CA8788875C2DA from managed side with 222 integrations.
09/20/21 03:29:16.897 PM [10372|6196] [info] Request ReJIT done for 6 methods
09/20/21 03:29:16.898 PM [10372|9972] [info] InitializeProfiler: Total integrations in profiler: 222
09/20/21 03:29:17.008 PM [10372|6196] [info] Request ReJIT done for 1 methods
09/20/21 03:29:17.284 PM [10372|1764] [info] *** CallTarget_RewriterCallback() Finished: System.Net.HttpWebRequest.BeginGetRequestStream() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.HttpWebRequest_BeginGetRequestStream_Integration, Arguments=2]
09/20/21 03:29:17.451 PM [10372|1764] [info] *** CallTarget_RewriterCallback() Finished: System.Net.WebRequest.GetResponseAsync() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.WebRequest_GetResponseAsync_Integration, Arguments=0]
09/20/21 03:29:17.470 PM [10372|1764] [info] *** CallTarget_RewriterCallback() Finished: System.Net.HttpWebRequest.BeginGetResponse() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.HttpWebRequest_BeginGetResponse_Integration, Arguments=2]
09/20/21 03:29:17.488 PM [10372|6196] [info] Request ReJIT done for 12 methods
09/20/21 03:29:17.519 PM [10372|6900] [info] *** CallTarget_RewriterCallback() Finished: System.Net.HttpWebRequest.EndGetResponse() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.HttpWebRequest_EndGetResponse_Integration, Arguments=1]
09/20/21 03:29:18.734 PM [10372|5632] [info] AssemblyLoadFinished: Datadog.Trace.dll v1.28.6.0 matched profiler version v1.28.6.0
09/20/21 03:29:18.734 PM [10372|5632] [info] AssemblyLoadFinished: Datadog.Trace.dll was not loaded domain-neutral
09/20/21 03:29:18.835 PM [10372|6196] [info] Request ReJIT done for 4 methods
09/20/21 03:29:18.861 PM [10372|6196] [info] Request ReJIT done for 1 methods
09/20/21 03:29:19.047 PM [10372|5632] [info] JITCompilationStarted: Startup hook registered in function_id=142890696 token=100664066 name=Xunit.DiaSessionWrapperHelper..ctor(), assembly_name=xunit.runner.utility.net452 app_domain_id=109966624 domain_neutral=0
09/20/21 03:29:19.095 PM [10372|5632] [info] ModuleLoadFinished: Datadog.Trace.ClrProfiler.Managed.Loader loaded into AppDomain 109966624 xunit.runner.utility.net452
09/20/21 03:29:19.398 PM [10372|5632] [info] AssemblyLoadFinished: Datadog.Trace.dll v1.28.6.0 matched profiler version v1.28.6.0
09/20/21 03:29:19.398 PM [10372|5632] [info] AssemblyLoadFinished: Datadog.Trace.dll was not loaded domain-neutral
09/20/21 03:29:19.432 PM [10372|6196] [info] Request ReJIT done for 6 methods
09/20/21 03:29:19.911 PM [10372|5632] [info] InitializeProfiler: received id: FFAFA5168C4F4718B40CA8788875C2DA from managed side with 222 integrations.
09/20/21 03:29:19.911 PM [10372|5632] [info] InitializeProfiler: Id already processed.
09/20/21 03:29:20.146 PM [10372|6196] [info] Request ReJIT done for 1 methods
09/20/21 03:29:20.517 PM [10372|10252] [info] *** CallTarget_RewriterCallback() Finished: System.Net.HttpWebRequest.BeginGetRequestStream() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.HttpWebRequest_BeginGetRequestStream_Integration, Arguments=2]
09/20/21 03:29:20.620 PM [10372|6196] [info] Request ReJIT done for 4 methods
09/20/21 03:29:20.803 PM [10372|10252] [info] *** CallTarget_RewriterCallback() Finished: System.Net.WebRequest.GetResponseAsync() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.WebRequest_GetResponseAsync_Integration, Arguments=0]
09/20/21 03:29:20.824 PM [10372|10252] [info] *** CallTarget_RewriterCallback() Finished: System.Net.HttpWebRequest.BeginGetResponse() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.HttpWebRequest_BeginGetResponse_Integration, Arguments=2]
09/20/21 03:29:20.955 PM [10372|6900] [info] *** CallTarget_RewriterCallback() Finished: System.Net.HttpWebRequest.EndGetResponse() [IsVoid=0, IsStatic=0, IntegrationType=Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.WebRequest.HttpWebRequest_EndGetResponse_Integration, Arguments=1]
09/20/21 03:29:21.080 PM [10372|5632] [info] AssemblyLoadFinished: Datadog.Trace.dll v1.28.6.0 matched profiler version v1.28.6.0
09/20/21 03:29:21.080 PM [10372|5632] [info] AssemblyLoadFinished: Datadog.Trace.dll was not loaded domain-neutral
09/20/21 03:29:21.140 PM [10372|6196] [info] Request ReJIT done for 12 methods
09/20/21 03:29:21.498 PM [10372|5632] [info] JITCompilationStarted: Startup hook registered in function_id=187586628 token=100664278 name=Xunit.AssemblyHelper..ctor(), assembly_name=xunit.runner.utility.net452 app_domain_id=162633248 domain_neutral=0
09/20/21 03:29:21.538 PM [10372|5632] [info] ModuleLoadFinished: Datadog.Trace.ClrProfiler.Managed.Loader loaded into AppDomain 162633248 Datadog.Trace.DuckTyping.Tests
09/20/21 03:29:21.834 PM [10372|5632] [info] AssemblyLoadFinished: Datadog.Trace.dll v1.28.6.0 matched profiler version v1.28.6.0
09/20/21 03:29:21.834 PM [10372|5632] [info] AssemblyLoadFinished: Datadog.Trace.dll was not loaded domain-neutral
09/20/21 03:29:21.862 PM [10372|6196] [info] Request ReJIT done for 6 methods
09/20/21 03:29:22.240 PM [10372|5632] [info] InitializeProfiler: received id: FFAFA5168C4F4718B40CA8788875C2DA from managed side with 222 integrations.
09/20/21 03:29:22.240 PM [10372|5632] [info] InitializeProfiler: Id already processed.
```

@DataDog/apm-dotnet

### CODE REVIEW

File: tracer/src/Datadog.Trace/ClrProfiler/InstrumentationDefinitions.cs [hunk 1]
```diff
@@ -9,9 +9,23 @@ internal static partial class InstrumentationDefinitions
...
         private static string assemblyFullName = typeof(InstrumentationDefinitions).Assembly.FullName;
 
-        internal static NativeCallTargetDefinition[] GetAllDefinitions()
+        internal static Payload GetAllDefinitions()
         {
-            return GetDefinitionsArray();
+            return new Payload
+            {
+                // Fixed Id for definitions payload (to avoid loading same integrations from multiple AppDomains)
+                DefinitionsId = "FFAFA5168C4F4718B40CA8788875C2DA",
```
Comment (zacharycmontoya) [inline tracer/src/Datadog.Trace/ClrProfiler/InstrumentationDefinitions.cs:17]: Will this ever change? Do you expect a scenario where we'd need to load definitions twice?

Comment (tonyredondo) [reply tracer/src/Datadog.Trace/ClrProfiler/InstrumentationDefinitions.cs:17]: I don't expect that scenario tbh. I think we should load the integrations only once (like we did with the json file). So I don't expect that Id to change.

**Review by zacharycmontoya — APPROVED** [review]

File: tracer/src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 2]
```diff
@@ -1091,11 +1091,20 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITInlining(FunctionID callerId, Function
...
+void CorProfiler::InitializeProfiler(WCHAR* id, CallTargetDefinition* items, int size)
 {
     auto _ = trace::Stats::Instance()->InitializeProfilerMeasure();
 
-    Logger::Info("InitializeProfiler: received from managed side: ", size, " integrations.");
+    WSTRING definitionsId = WSTRING(id);
+    Logger::Info("InitializeProfiler: received id: ", definitionsId, " from managed side with ", size,
+                 " integrations.");
+
+    if (definitions_ids_.find(definitionsId) != definitions_ids_.end())
```
Comment (gleocadie) [inline tracer/src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:1102]: shouldn't the check in the set be done covered by a lock ?

**Review by gleocadie — COMMENTED** [review]
One question

File: tracer/src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 2 above)
Comment (tonyredondo) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:1102]: you're right, fixing it.

Comment (tonyredondo) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:1104]: done in 3925992

**Review by andrewlock — APPROVED** [review]
LGTM AFAICT! ** shakes fist at C++ **

**Pull request merged**